File: 2-27-24/TreeDiameter.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private int maxDiameter;

    public int diameterOfBinaryTree(TreeNode root) {
        maxDiameter = 0;

        maxDepth(root);

        return maxDiameter;
    }

    private int maxDepth(TreeNode node) {
        if(node == null) {
            return 0;
        }
        
        int leftDepth = 0;
        int rightDepth = 0;
        
        leftDepth = maxDepth(node.left);
        rightDepth = maxDepth(node.right);
        

        int diameter = leftDepth + rightDepth;
        if(diameter > maxDiameter) {
            maxDiameter = diameter;
        }

        return (leftDepth > rightDepth) ? (leftDepth + 1) : (rightDepth + 1);
    }
}

File: 2-27-24/TreeDiameter.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        def maxDepth(node, maxDiameter) -> int:
            if not node:
                return 0

            leftDepth = maxDepth(node.left, maxDiameter)
            rightDepth = maxDepth(node.right, maxDiameter)

            if leftDepth + rightDepth > maxDiameter[0]:
                maxDiameter[0] = leftDepth + rightDepth

            return leftDepth + 1 if leftDepth > rightDepth else rightDepth + 1

        maxDiameter = [0]
        maxDepth(root, maxDiameter)
        return maxDiameter[0]
